File: src/ros2_demo_cpp/src/hw_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/hardware_status.hpp"

class HardwareStatusPublisherNode : public rclcpp::Node
{
public:
    HardwareStatusPublisherNode() : Node("hardware_publisher")
    {
        _pub = this->create_publisher<my_robot_interfaces::msg::HardwareStatus>("hardware_status", 10);
        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&HardwareStatusPublisherNode::publishHardWareStatus, this));
        RCLCPP_INFO(this->get_logger(), "Hardware status has been started");
    }
private:
    void publishHardWareStatus(){
        auto msg = my_robot_interfaces::msg::HardwareStatus();
        msg.temperature = 14;
        msg.are_motors_ready = true;
        msg.debug_message = "It's too cold in Nghe An (Jan 17 2023)";
        _pub->publish(msg);
    }

    rclcpp::Publisher<my_robot_interfaces::msg::HardwareStatus>::SharedPtr _pub;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HardwareStatusPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
}

File: src/ros2_demo_py/ros2_demo_py/hw_status_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from my_robot_interfaces.msg import HardwareStatus

# create class to excute
class HardwareStatusPublisherNode(Node):
    def __init__(self):
        super().__init__("hardware_status_publisher")
        self.hw_status_publisher = self.create_publisher(HardwareStatus, "hardware_status", 10)
        self.timer = self.create_timer(1.0, self.hardware_hw_status)
        self.get_logger().info("Hardware status publisher has been started")
    def hardware_hw_status(self):
        msg = HardwareStatus()
        msg.temperature = 45
        msg.are_motors_ready = True
        msg.debug_message = "no things special here"
        self.hw_status_publisher.publish(msg)

    def timer_callback(self):
        self.counter_ += 1
        self.get_logger().info("TIMES " + str(self.counter_))

def main(args = None):
    rclpy.init(args=args)
    node = HardwareStatusPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()